src/CentrifugeRouter.sol

```diff
@@ -2,97 +2,197 @@
+        require(asset != address(0), "CentrifugeRouter/unknown-vault");
+        SafeTransferLib.safeTransferFrom(asset, owner, escrow, amount);
+
+        lockedRequests[controller][vault] += amount;
+        emit LockDepositRequest(vault, controller, owner, _initiator, amount);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function unlockDepositRequest(address vault) external {
-        uint256 lockedRequest = lockedRequests[msg.sender][vault];
+    function unlockDepositRequest(address vault) external protected {
+        uint256 lockedRequest = lockedRequests[_initiator][vault];
         require(lockedRequest > 0, "CentrifugeRouter/user-has-no-locked-balance");
-        lockedRequests[msg.sender][vault] = 0;
-        SafeTransferLib.safeTransfer(IERC7540Vault(vault).asset(), msg.sender, lockedRequest);
-        emit UnlockDepositRequest(vault, msg.sender);
+        lockedRequests[_initiator][vault] = 0;
+
+        address asset = poolManager.vaultToAsset(vault);
+
+        require(asset != address(0), "CentrifugeRouter/unknown-vault");
+        EscrowLike(escrow).approve(asset, address(this), lockedRequest);
+
+        SafeTransferLib.safeTransferFrom(asset, escrow, _initiator, lockedRequest);
+        emit UnlockDepositRequest(vault, _initiator);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function executeLockedDepositRequest(address vault, address user) external {
-        uint256 lockedRequest = lockedRequests[user][vault];
-        require(lockedRequest > 0, "CentrifugeRouter/user-has-no-balance");
-        lockedRequests[user][vault] = 0;
-        IERC7540Vault(vault).requestDeposit(lockedRequest, user, address(this));
-        emit ExecuteLockedDepositRequest(vault, user);
+    function executeLockedDepositRequest(address vault, address controller) external protected {
+        uint256 lockedRequest = lockedRequests[controller][vault];
+        require(lockedRequest > 0, "CentrifugeRouter/controller-has-no-balance");
+        lockedRequests[controller][vault] = 0;
+
+        address asset = poolManager.vaultToAsset(vault);
+
+        require(asset != address(0), "CentrifugeRouter/unknown-vault");
+        EscrowLike(escrow).approve(asset, address(this), lockedRequest);
+        SafeTransferLib.safeTransferFrom(asset, escrow, address(this), lockedRequest);
+
+        _approveMax(asset, vault);
+        IERC7540Vault(vault).requestDeposit(lockedRequest, controller, address(this));
+        emit ExecuteLockedDepositRequest(vault, controller, _initiator);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimDeposit(address vault, address user) external {
-        uint256 maxDeposit = IERC7540Vault(vault).maxDeposit(user);
-        IERC7540Vault(vault).deposit(maxDeposit, user, user);
+    function claimDeposit(address vault, address receiver, address controller) external protected {
+        uint256 maxDeposit = IERC7540Vault(vault).maxDeposit(controller);
```

Comment:
we have to make sure that the controller is the _initiator, otherwise anybody can pass a random receiver for a controller that has a balance and steal the funds

Reply:
Wow, I'm not sure how I overlooked that, but you're of course right. We can just remove the `controller` praameter here

Reply:
Fixed in the latest commit. You can now either claim on behalf of someone else, to their address, or on behalf of yourself to any address.